지도의 스타일을 변경 가능한 util함수 및 각 feature별로 변경을 위한 함수 작성

qkrdmstlr3 opened the pull request:
## 상세사항
- 지도의 스타일을 변경하기 위해서 util폴더에 applyStyle파일 안에 각 속성별로 다섯가지 함수를 만듦
- useStyle훅에서 dispatch전달 전에 함수(map-styling)를 실행하고, 필요한 요소들을 전달.
- map-styling에서 전달받은 값들을 이용해서, 분류에 따라 applyStyle의 함수를 실행 ( 분류는 구현해야함 )

## 기타사항
- input range로 스타일을 바꾸는 부분이, 원래 상태로 돌아가도 체크가 해제되지 않는 이슈가 발생해서 string으로 처리를 해주었더니 정상적으로 동작함
- 지금 보니 color를 hsl로 바꿀 때 s, l값을 넣어주면 좀 더 짧게 구현할 수 있었겠네요;;

qkrdmstlr3 committed:
[fix] visibility의 value값들 수정

- 좀 더 편하게 처리를 하기 위해서 value값들을 알맞게 수정

qkrdmstlr3 committed:
[fix] 이전, 현재 상태 비교할 때 에러 수정

- 이전과 현재 스타일의 상태를 비교할 때 자료형이 맞지않아 원하는대로 작동하지 않음
- string으로 처리를 해주어서 비교함

qkrdmstlr3 committed:
[feat] 지도에 스타일을 적용할 수 있는 모듈을 만듦

- 지도에 각각의 스타일들을 적용할 수 있게 여러 함수들을 모듈로 만듦
- 각각의 타입(symbol line polygon)마다 서로 다른 값들을 인자로 주어야 함

qkrdmstlr3 committed:
[feat] 스타일 변경 후 각 상태들을 바꿀 수 있게 utils파일에 feature별로 나누어 놓음

- 스타일 변경 후 dispatch를 보내기 이전에 지도의 상태를 변경가능하게 feature단위로 파일을 만들어 놓고 useStyle훅에서 불러서 가져옴
- 각 파일들의 내부에서 레이어 변경 및 기타 다른 작업들을 진행하면 좋을듯

qkrdmstlr3 committed:
[refactor] 넘겨줄 때 인자로 style값을 추가

- 넘겨줄 때 스타일 적용이 가능하도록 style을 추가함( 사실 빼먹었었음 )

qkrdmstlr3 committed:
[fix] hsl색상 변경시에 부족한 부분이 있어서 수정

- hsl색상 변경시 부족한 부분이 많아서 수정을 함
- 1. s와 l엔 %를 붙여주어야 함
- 2. hex를 rgb로 변경 시 16진수에서 바꾸어 주어야함

qkrdmstlr3 committed:
[feat] applyStyle인자들 서순 변경

- applyStyle의 매개변수들의 순서를 변경하였다.
- 나중에 object로 바꿀 필요 있을 듯

sayhs30 commented on src/utils/colorFormat.ts:
```diff
@@ -0,0 +1,42 @@
+  l: number;
+}
+
+export function hexToHSL(color: string): HexToHSLType {
```
타입도 추가하면, 코드보는데 더 편리할거같습니다.

qkrdmstlr3 replied:
내부 변수에 타입을 추가하는 것을 말씀하시는게 맞을까요??
그리고 그냥 궁금한 점인데 모든 변수들에 대해서 타입정의를 해주는 것이 좋을까요??

sayhs30 replied:
처음에는 매우 매우 번거롭지만, 그에 따른 이득도 있습니다.

정적 타입 검사를 통해 오류를 미리 찾을수도 있고, 가독성도 높아지고, 코딩 할 때도 생산성을 높일 수 있습니다 ㅎㅎ

상황에 맞게 사용 여부는 결정하시면 되지만, 저는 추천 합니다 ㅎ

sayhs30 commented on src/hooks/sidebar/useStyleType.ts:
```diff
@@ -30,18 +32,29 @@ function useStyleType({
 }: UseStyleTypeProps): UseStyleHookType {
   const dispatch = useDispatch();
 
+  const map = useSelector<RootState>((state) => state.map.map) as mapboxgl.Map;
```
추가로 구현된 hook들을 보니, 
어떤 state는 props로 전달받고, 어떤 항목은 내부에서 store의 값을 참조하고 있는데요, 이 경우 재사용이 어려울 거같은데, 어떤 의도하신 목적이있나요?

프로젝트 구조에서 어떤 경우에 store와 hook를 사용하기로 했는지 궁금합니다~!

qkrdmstlr3 replied:
첫 번째 질문에 대한 답을 먼저 드리자면, map이라는 것을 코드의 여러 부분에서 사용해야 할 것 같아서(useMap, useLowerButtons, useStyleType), 전역으로 관리를 하자는 이야기가 나왔고, 리덕스를 사용하면 관리가 좀 더 쉬워질 것 같아서 그렇게 결정하였습니다.
두 번째 질문은 일단 hooks/common에 들어있는 hook들은 재사용성이 될 것이라고 보고 만들었고, 나머지 hook들은 재사용이 목적이 아닌, view와 model을 분리하기 위해서 만들었습니다. 초기에 저희가 container - presentational패턴을 사용했었는데 파일 구조가 얽히면서, presentation을 메인으로 사용하고 container역할을 hook으로 분리해서 사용하는 것이 좋지 않을까라는 생각으로 리팩토링을 하였었습니다.
혹시라도 제가 질문을 잘못 이해했거나, 제가 잘못 알고 있는 부분이 있다면 말씀해주시면 감사하겠습니다!

gitgitWi replied:
멘토님 첫번째 질문에 대해서 저도 이게 맞나 싶었던 부분인데,
`map` 자체는 저장된 상태를 받는 컴포넌트니까 store 에서 분리하고, 
각 `Styler`에서 변경하는 상태만 이동할 수 있게 하는 방법을 좀더 고민해보면 좋을 것 같아요!

Eunsol0410 replied:
은식님 답변에 더하자면 hook에 영향을 받지 않는 View 컴포넌트만 재사용 되는 경우는 드물어서 이렇게 결정 하게 되었습니다. 간혹있는 재사용이 요구되는 View(Presentational) 컴포넌트의 경우 hook과 분리된 독립적 컴포넌트로 작성하였습니다.

qkrdmstlr3 replied:
저도 아까 느꼈던 건데, style을 바꾸려해서 map을 불러와야했는데 useStyleType훅에서 스타일을 바꾸는 함수까지 map을 전달전달해야 하는 불편함이 있었습니다. 분리를 하면은 이런 부분이 좀 더 개선될 수 있을 것 같네요!

sayhs30 replied:
@qkrdmstlr3 @gitgitWi @Eunsol0410 

답변감사합니다! 

hook/store 사용 시 어떤 기준을 두고 사용하시는지 궁금해서 여쭤봤습니다. 많이 고민 하신거같네요 ㅎㅎ 

@gitgitWi 님, 말씀하신 방법도 좋은거같습니다! 다른 분들과 고민해보시면 좋을거같습니다.

 Map 컴포넌트/map 객체/스토어에 저장되는 state(예: 스타일)들의 각각의 역할과 서로 간의 관계를 생각해보면
뭐는 어디서 담당하고, 각 역할간 인터페이스는 어떻게 맞추는게 좋을지 결정하는데 도움이 될거같습니다.

qkrdmstlr3 commented on src/components/Sidebar/SidebarContentMore/VisibilityStyle.tsx:
```diff
@@ -55,8 +55,8 @@ function VisibilityStyle({
-    { title: '보임', value: 'visibility' },
-    { title: '숨김', value: 'invisibility' },
+    { title: '보임', value: 'visible' },
+    { title: '숨김', value: 'none' },
```
아이구야 enum은 뭔가 아직도 낯선 친구여서 잘 까먹네요;;

qkrdmstlr3 committed:
[feat] 메인화면에 사용할 디자인 및 컴포넌트 생성

- 메인화면에 사용할 컴포넌트를 canvas를 이용해서 구상해보았다
- 필요한 이미지 파일들 추가
- 기존과 동일하게 Canvas컴포넌트 생성 후 useCanvas hook으로 로직을 가져왔다

gitgitWi commented on src/hooks/sidebar/useStyleType.ts:
```diff
@@ -1,3 +1,4 @@
+import mapboxgl from 'mapbox-gl';
```
오로지 타입 강제만을 위해서 불러오는 거라면, common type 쪽에 타입 선언을 하나 해주는게 좋을 것 같아요

qkrdmstlr3 replied:
이게 되게 애매하더라구요. common쪽에 빼주어도  export type MapType = mapboxgl.Map 이렇게 한줄이면 처리될텐데, import가 두번 쓰이는 느낌? 물론 import경로로 보았을 때 좀 더 명시적인 장점도 있을 것 같은데... 어떻게 생각하세요???

gitgitWi replied:
import가 어떻게 작동하는지는 잘 모르지만, 
느낌상 라이브러리 하나를 전부 불러오면 너무 무거워 질 거 같아서 타입만 하나 빼서 재사용하면 될 것 같아서 그랬어요 ㅋㅋㅋㅋ
이부분 좀더 알아볼게요...ㅋㅋ

gitgitWi reviewed (approved):
고생많으셨습니다!
이제 여기에 리덕스를 추가하면 되는 거군여!

Eunsol0410 commented on src/store/style/compareStyle.ts:
```diff
@@ -6,7 +6,9 @@ export function checkStyleIsChanged(targetStyle: StyleType): boolean {
-    (key) => key === 'isChanged' || defaultStyle[key] === targetStyle[key]
+    (key) =>
+      key === 'isChanged' ||
+      String(defaultStyle[key]) === String(targetStyle[key])
```
`String` 타입 에러가 발생해서 처리한걸까요?

gitgitWi replied:
린터 때매 줄바꿈인 줄 알았는데 String이 있었네요.. 저도 궁금합니다

qkrdmstlr3 replied:
기존에 코드를 돌려보니까 range를 바꾸는 부분에 있어서는 체크표시가 원상태로 복귀해도 해제가 안되더군요. 그래서 문제점을 찾던 중 저 부분이 문제라는 것을 발견했고, 시간 상 응급조치를 일단 해둔것입니다!

qkrdmstlr3 replied:
혹시 다른 팀에서는 괜찮나요? 제가 뭘 건드려서 그런가..?

GodDrinkTeJAVA replied:
리뷰 늦어서 죄송합니다...!

저도 테스트해볼 때 왜 저러지 했었는데 

기본 설정값은 Number로 되어있는데

input type="range"에서 onChange로 들어오는 값은 기본적으로 string(ex. "6")이 들어오기 때문이 아닌가 싶습니다. 저희가 onChange 단에서 미리 Number로 변환 처리를 해줘야 할 거 같습니다

yyjjjj replied:
이 부분을 처리할때 준희님 말씀처럼 드래그바의 onChange 단에서 변환처리해주거나, 지금 현재 처리되고 있는 것처럼 요소들에 대하여 한꺼번에 처리될때 드래그바 요소 외의 것들이 string 이라는 점을 이용하여 비교시 타입을 통일해버리는 방법이 있다고 생각합니다.
단순히 비교처리만 하면 된다고 생각하여 후자를 택했었는데, 지금은 기존의 타입을 해치는 방법이라 좋지못하다는 생각이 드네요. 준희님 말씀처럼 리팩토링시 onChange 단에서의 변환을 고려해보는것도 좋을것 같습니다.

Eunsol0410 commented on src/store/style/properties.ts:
```diff
@@ -3,8 +3,8 @@ import { StyleType, ElementType, FeatureType } from '../common/type';
   visibility: 'inherit',
-  color: '#000000',
-  weight: 4,
+  color: '#55bf40',
```
다시 생각해보니... default 스타일과 실제 레이블에 설정된 스타일 값의 싱크가 안맞아도 괜찮을까요?

qkrdmstlr3 replied:
저도 연정님이랑 그 얘기를 한 부분인데, 구글과 비교를 해보면 채도 변경시에 색상은 변하지않고 그대로 있더군요. 그래서 일단은 색상 / 채도 / 밝기 변경시 서로가 서로에게 영향을 주지는 말자로 결론을 내리긴 했었습니다!

Eunsol0410 reviewed (approved):
수고하셨습니다! 내일 스타일이 바뀌는 모습을 보면 정말 뿌듯할 것 같네요 ^^

qkrdmstlr3 committed:
[feat] [#79] mapbox-poi분리

- mapbox-poi를 사이트를 참조해서 분리해놓음
- 전체적인 형식은 기존 poi형식을 따름
